docs/blueprints/forgotten-deletes.md

```diff
@@ -0,0 +1,118 @@
+# Forgotten Deletes
+
+## Abstract
+
+In the event of a node being audited for a piece which the satellite previously asked it to delete via delete pieces request, the node will present a signed message proving the legitimacy of the deletion. The node is not penalized and the object is deleted. We eliminate the potential of forgotten garabage collection deletes entirely.
+
+## Background
+
+In the event of a DB restoration from backup, it may be possible for the satellite to restore pointers for objects which were deleted, or revert pointers to a previous state containing nodes which no longer have those pieces. In this situation, the satellite would audit nodes for pieces which they were told to delete which may result in unfair disqualifications.
+
+## Design
+
+The solution to this problem is twofold, as there are two ways in which pieces are deleted: delete pieces requests and garbage collection.
+
+### Delete pieces requests
+
+Delete pieces requests are issued when a user deletes an object. The metainfo service deletes the object's pointers and then sends delete pieces requests to each of the reliable nodes which held pieces. The rest of the nodes' pieces will be eventually deleted by garbage collection. 
+
+To prove to the satellite that a piece was deleted legitimately, we will have it sign the delete pieces requests. The node will present these to the satellite in the event of an audit for a deleted piece.
+
+```
+message DeletePiecesRequest {
+    repeated bytes piece_ids = 1 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+    bytes satellite_signature = 2;
+}
+```
+
+The storage nodes will serialize and save these signed messages into a new database table.
+
+```sql
+CREATE TABLE delete_pieces_requests (
+    id INTEGER PRIMARY KEY,
+    data BYTEA,
+    created_at TIMESTAMP,
+)
+```
+
+The storage node now has proof that the satellite requested these pieces to be deleted. Now we need the ability to access this signed message if given a particular piece ID.
+
+These signed messages may contain multiple piece IDs. We need to be able to reference this row if given any one piece ID contained within the message.
+
+To do this we create another table to hold each deleted piece with a foreign key which references the corresponding row in delete_pieces_requests to retrieve the signed delete request.
+
+```sql
+CREATE TABLE deleted_pieces (
+    id BYTEA,
+    signed_message INTEGER,
+    FOREIGN KEY(signed_message) REFERENCES deleted_pieces_requests(id)
+)
```

Comment:
oh now i'm very worried. this seems like quite a bit of overhead on every delete. we should avoid doing as much as we can in normal hotpath of deletes imo

Reply:
Paul had the same concern here and suggested we implement a file store for handling these. I was getting kind of stuck trying to design it and thought it might be better to just get the document out and discuss that aspect during the review

Reply:
It is a lot of overhead - is there any way we could combine the two tables into one? How often would a storagenode actually get a delete request with multiple pieces in it? I feel like the majority would just be to delete a single piece, in which case it is a lot of trouble to do two separate table updates.

@jtolio I am having a hard time thinking of how we could accomplish storing this proof without a solution that uses the database, but it would be nice if we could.
Off the top of my head, I wonder if it would be possible to store some file at the location of the old piece with the signed delete request that the storagenode could search for - something that would be distinguished from a normal data blob, but that would be a clear indicator of "the piece at this path is deleted and here is proof". Any initial thoughts on that idea?

Reply:
> How often would a storagenode actually get a delete request with multiple pieces in it?

Actually, this happens much more often than you'd think, though we should indeed measure it. Satellites have a batching system that batch up incoming storage node delete requests (across different user requests) and send them out together.

> Off the top of my head, I wonder if it would be possible to store some file at the location of the old piece with the signed delete request that the storagenode could search for - something that would be distinguished from a normal data blob, but that would be a clear indicator of "the piece at this path is deleted and here is proof"

Using the filesystem as a key/value database isn't a bad idea! That could work. 

One terrible idea is they all just get stored in completely unsorted order but in identifiable buckets (deleted pieces that start with a1..., for example), and then in the rare case that a signed satellite request asking for a piece that the storage node doesn't have actually shows up, the storage node can sequentially scan through that entire bucket. Responding slow after scanning a large file is potentially better than just failing the audit, but it might be untenable, I don't know. I guess it depends how often deletes happen and how large these files would be.

Reply:
I don't think that is a terrible idea. In theory, a node should get audited for a piece that the satellite asked to delete extremely rarely if ever, so the the performance downside of iterating over a bucket in that case probably wouldn't be so bad all things considered. And we could tweak bucket size depending on what we want.

I'm not sure how to weigh it against the solution I proposed above.

Reply:
yeah i think they might be both workable. we probably need some numbers of what's currently happening with deletes to be able to say

Reply:
In either case we should have a chore that iterates over the deleted file proofs and removes any that are far enough in the past.

Reply:
Yeah, my suggestion was just to add piece IDs and the corresponding delete orders to a flat text file, and just rotate the file daily, deleting ones that are old enough. But yeah definitely we could "shard" that file by piece ID prefix to whatever degree is necessary.